File: Algorithms/3. Longest Substring Without Repeating Characters/python/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        # mp stores the current index of a character
        mp = {}

        i = 0
        # try to extend the range [i, j]
        for j, char in enumerate(s):
            if char in mp:
                i = max(mp[char], i)

            ans = max(ans, j - i + 1)
            mp[char] = j + 1

        return ans

File: Algorithms/3. Longest Substring Without Repeating Characters/swift/solution.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var ans = 0
        var mp: [Character: Int] = [:]

        var i = 0
        for (j, char) in s.enumerated() {
            if let val = mp[char] {
                i = max(i, val)
            }

            ans = max(ans, j - i + 1)
            mp[char] = j + 1
        }
        return ans
    }
}
